src/rgw/rgw_role.cc

```diff
@@ -0,0 +1,275 @@
+}
+
+void RGWRole::decode_json(JSONObj *obj)
+{
+  JSONDecoder::decode_json("id", id, obj);
+  JSONDecoder::decode_json("name", name, obj);
+  JSONDecoder::decode_json("path", path, obj);
+  JSONDecoder::decode_json("create_date", creation_date, obj);
+  JSONDecoder::decode_json("assume_role_policy_document", trust_policy, obj);
+}
+
+int RGWRole::read_id(const string& role_name, string& role_id)
+{
+  auto& pool = store->get_zone_params().roles_pool;
+  string oid = get_names_oid_prefix() + role_name;
+  bufferlist bl;
+  RGWObjectCtx obj_ctx(store);
+
+  int ret = rgw_get_system_obj(store, obj_ctx, pool, oid, bl, NULL, NULL);
+  if (ret < 0) {
+    return ret;
+  }
+
+  RGWNameToId nameToId;
+  try {
+    bufferlist::iterator iter = bl.begin();
+    ::decode(nameToId, iter);
+  } catch (buffer::error& err) {
+    ldout(cct, 0) << "ERROR: failed to decode role from pool: " << pool.name << ": "
+                  << role_name << dendl;
+    return -EIO;
+  }
+  role_id = nameToId.obj_id;
+  return 0;
+}
+
+int RGWRole::read_info()
+{
+  auto& pool = store->get_zone_params().roles_pool;
+  string oid = get_info_oid_prefix() + id;
+  bufferlist bl;
+  RGWObjectCtx obj_ctx(store);
+
+  int ret = rgw_get_system_obj(store, obj_ctx, pool, oid, bl, NULL, NULL);
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: failed reading role info from pool: " << pool.name <<
+                  ": " << id << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  try {
+    bufferlist::iterator iter = bl.begin();
+    ::decode(*this, iter);
+  } catch (buffer::error& err) {
+    ldout(cct, 0) << "ERROR: failed to decode role info from pool: " << pool.name <<
+                  ": " << id << dendl;
+    return -EIO;
+  }
+
+  return 0;
+}
+
+int RGWRole::read_name()
+{
+  auto& pool = store->get_zone_params().roles_pool;
+  string oid = get_names_oid_prefix() + name;
+  bufferlist bl;
+  RGWObjectCtx obj_ctx(store);
+
+  int ret = rgw_get_system_obj(store, obj_ctx, pool, oid, bl, NULL, NULL);
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: failed reading role name from pool: " << pool.name << ": "
+                  << name << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  RGWNameToId nameToId;
+  try {
+    bufferlist::iterator iter = bl.begin();
+    ::decode(nameToId, iter);
+  } catch (buffer::error& err) {
+    ldout(cct, 0) << "ERROR: failed to decode role name from pool: " << pool.name << ": "
+                  << name << dendl;
+    return -EIO;
+  }
+  id = nameToId.obj_id;
+  return 0;
+}
+
+int RGWRole::update_trust_policy(string& trust_policy)
+{
+  auto& pool = store->get_zone_params().roles_pool;
+
+  int ret = get();
+  if (ret < 0)
+  {
+    return ret;
+  }
+
+  this->trust_policy = trust_policy;
+
+  ret = store_info(false);
+  if (ret < 0)
+  {
+    ldout(cct, 0) << "ERROR:  updating role info in pool: " << pool.name << ": "
+                  << id << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+}
```

Comment:
@cbodley: Yeah initially, i thought of breaking the logic as you have suggested, but if we implement the AWS API, then the same code (that will be there in rgw-admin.cc), will be replicated in the class which will have the corresponding method for role modify API. Thats the reason why I put everything under one interface in RGWRole class. Let me know what you think.